lib/ts/recipe/thirdparty/recipeImplementation.ts

```diff
@@ -192,29 +159,37 @@ export default function getRecipeImplementation(recipeId: string, appInfo: Norma
                   status: "NO_EMAIL_GIVEN_BY_PROVIDER";
                   fetchResponse: Response;
               }
-            | {
-                  status: "FIELD_ERROR";
-                  error: string;
-                  fetchResponse: Response;
-              }
         > {
-            const stateFromStorage = this.getOAuthState({
+            const stateFromStorage = this.getStateAndOtherInfoFromStorage<{}>({
                 userContext: input.userContext,
                 config: input.config,
-            }).state;
-
-            const code = getQueryParams("code");
+            });
 
             const stateFromQueryParams = getQueryParams("state");
 
             if (
-                getQueryParams("error") !== null ||
                 stateFromStorage === undefined ||
-                stateFromStorage.thirdPartyId !== input.thirdPartyProviderId ||
-                stateFromStorage.state !== stateFromQueryParams ||
-                code === null
+                !this.verifyStateFromOAuthProvider({
+                    stateFromProvider: stateFromQueryParams,
+                    stateFromStorage: stateFromStorage.stateForAuthProvider,
+                    config: input.config,
+                    userContext: input.userContext,
+                })
             ) {
-                throw new STGeneralError("");
+                throw new Error("Invalid 'state' recieved from provider");
+            }
+
+            const code = input.authCode === undefined ? getQueryParams("code") : input.authCode;
+
+            if (code === undefined) {
+                throw new Error(
+                    "There is no 'code' present in query params and no 'authCode' was provided when calling signInUp"
+                );
+            }
+
+            if (getQueryParams("error") !== undefined) {
+                // TODO NEMI: This should have a better message
+                throw new Error("Something went Wrong");
             }
```

Comment:
Change to `"Something went rrong"`, or see if provider is sending a proper error message, then send that (research based on OAuth 2.0 spec)